api/ecosystem_test.go

```diff
@@ -3,15 +3,192 @@ package api
 // ecosystem_test.go provides tooling and tests for whole-ecosystem testing,
 // consisting of multiple full, non-state-sharing nodes connected in various
 // arrangements and performing various full-ecosystem tasks.
+//
+// To the absolute greatest extent possible, nodes are queried and updated
+// exclusively through the API.
 
 import (
 	"errors"
+	"net/url"
 	"testing"
 	"time"
 
 	"github.com/NebulousLabs/Sia/types"
 )
 
+// addStorageToAllHosts adds a storage folder with a bunch of storage to each
+// host.
+func addStorageToAllHosts(sts []*serverTester) error {
+	for _, st := range sts {
+		values := url.Values{}
+		values.Set("path", st.dir)
+		values.Set("size", "1048576")
+		err := st.stdPostAPI("/host/storage/folders/add", values)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
```

Comment:
I added a ton of ecosystem helper functions which make it easier to set up and command fleets of hosts and renters. This was needed for the current testing, as the renter now has a lot more parallelism and I needed to make sure that it could handle a lot of conditions. Notably missing from the current test suite are situations where hosts go offline.